st-embroider/2023/2023-03-07.md

```diff
@@ -0,0 +1,164 @@
+
+- EF: merged
+- EF: open PR on ember-auto-import that will start passing when we get a stable on embroider PR #571 on ember-auto-import for regression
+- https://github.com/ef4/ember-auto-import/pull/571
+
+
+
+### clarify services DI and resolution (and interactions with engines and addons)<@void_mAlex>:
+Given a structure similar to:
+
+sharedAddon
+  -addon
+	-services
+  	  shared
+  -app
+	-services
+  	  shared
+
+projectXengine/addon -depends on sharedAddon
+  -addon
+	-components
+  	  comp1 (injects @service internalService AND @service shared)
+	-services
+  	  internalService
+  -app
+	-components
+  	  comp1 <- component re-export
+	-services
+
+consumer -depends on sharedAddon and projectXengine/addon
+  -app
+	-components
+  	  widget <- uses comp1 from projectXengine/addon
+	-services
+  -tests
+	-integration
+  	  -components
+    	    widget <- error from missing service
+
+- V1 addon internal service staticAddonTrees = true
+  - Legacy build forces re-export of all services and given a naming collision in host and addon it would prefer lower service embroider seems to take the host
+  - Q: should internally used service be re-exported? Guessing the answer will be yes as without re-export under embroider it fails to find the service
+- V1 engine same as above - is there a difference from a plain addon and should users expect there to be one; does embroider do anything special for engines
+- Currently DI seems broken if consumer app injects the shared service from the addon and the engine/addon also injects it, they get different instances; not sure if that is intended behaviour or not
+- Expected behaviour when a service injects another service and that service is passed to an engine should engine have access to the injected service?
+
+- EF: we need to codify the old behaviour, not a lot of this was speced out
+- the best thing that you can do is to create a reproduction for situations that concern you - potentially in the engines test - `embroider/tests/scenarios/engines-test`
+- A: you can have a situation where you get a new instance of a service in an addon vs the consuming app
+- EF: you're right that this is relevant to the current work on resolver rules
+- A: when you have name clashes of services (api service in the engine and consumer app with api service)
+- it seems that in embroider you will always inject the consumer's api service when the engine requests it.
+- if you want an internal service to an engine are you expected to export it? 
+- is it a limitation of the DI system that you can't have isolated services
+- EF: my understanding is that engines use nested containers
+- let's get clear what the classic behaviour is and reflect it in the tests
+- we're at the point where we shouldn't be recommending engines (probably)
+- there are better ways to isolate developers from each other
```

Comment:
@ef4 sorry for post-meeting callout, are you talking about https://single-spa.js.org ?
as we are using engines for actual micro-frontends where services shared only for the purpose of sharing state and never services/components being re-exported from the engine (IMO I think this is just wrong use of engines called out by @void_mAlex, unless I'm missing smth)

Reply:
@SergeAstapov I believe the other effort would be an exploration that is attempted into micro-front ends as federated modules
engines work under embroider quite well, what I was bringing up were some questions around services and how they behave in the stricter modes especially where I was trying to identify discrepancies from classic builds (and if they exist figure out how to address them)
I'm using engines for separating standalone applications; there is a lot of context lost around the question which involved sharing a service from a normal addon into both an engine and a separate consuming app at the same time where the app also hosts the engine

Reply:
@void-mAlex thank you for the detailed explanation!